File: ingest/src/main/java/com/microsoft/azure/kusto/ingest/ColumnMapping.java
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

package com.microsoft.azure.kusto.ingest;

import org.apache.commons.lang3.StringUtils;

import java.io.Serializable;
import java.util.HashMap;
import java.util.Map;

/**
 * Data class that describes the Mapping setting for .ingest command
 */
public class ColumnMapping implements Serializable {
    private final String columnName;
    private final String columnType;
    private final Map<String, String> properties;

    public ColumnMapping(String columnName, String cslDataType) {
        this(columnName, cslDataType, new HashMap<>());
    }

    public ColumnMapping(String columnName, String cslDataType, Map<String, String> properties) {
        this.columnName = columnName;
        this.columnType = cslDataType;
        this.properties = properties;
    }

    public ColumnMapping(ColumnMapping other) {
        columnName = other.columnName;
        columnType = other.columnType;
        properties = new HashMap<>(other.properties);
    }

    public String getColumnName() {
        return columnName;
    }

    public String getColumnType() {
        return columnType;
    }

    public Map<String, String> getProperties() {
        return properties;
    }

    public void setPath(String path) {
        properties.put(MappingConsts.PATH.getName(), path);
    }

    public String getPath() {
        return properties.get(MappingConsts.PATH.getName());
    }

    public void setTransform(TransformationMethod transform) {
        properties.put(MappingConsts.TRANSFORMATION_METHOD.getName(), transform.name());
    }

    public TransformationMethod getTransform() {
        String transform = properties.get(MappingConsts.TRANSFORMATION_METHOD.getName());
        return StringUtils.isEmpty(transform) ? null : TransformationMethod.valueOf(transform);
    }

    public void setOrdinal(Integer ordinal) {
        properties.put(MappingConsts.ORDINAL.getName(), String.valueOf(ordinal));
    }

    Integer getOrdinal() {
        String ordinal = properties.get(MappingConsts.ORDINAL.getName());
        return StringUtils.isEmpty(ordinal) ? null : Integer.valueOf(ordinal);
    }

    public void setConstantValue(String constValue) {
        properties.put(MappingConsts.CONST_VALUE.getName(), constValue);
    }

    public String getConstantValue() {
        return properties.get(MappingConsts.CONST_VALUE.getName());
    }

    public void setField(String field) {
        properties.put(MappingConsts.FIELD_NAME.getName(), field);
    }

    String setField() {
        return properties.get(MappingConsts.FIELD_NAME.getName());
    }

    public void setColumns(String columns) {
        properties.put(MappingConsts.COLUMNS.getName(), columns);
    }

    public String getColumns() {
        return properties.get(MappingConsts.COLUMNS.getName());
    }

    public void setStorageDataType(String dataType) {
        properties.put(MappingConsts.STORAGE_DATA_TYPE.getName(), dataType);
    }

    public String getStorageDataType() {
        return properties.get(MappingConsts.STORAGE_DATA_TYPE.getName());
    }

    public boolean isValid(IngestionMapping.IngestionMappingKind mappingKind)
    {
        switch (mappingKind)
        {
            case Csv:
            case SStream:
                return !StringUtils.isEmpty(this.columnName);
            case Json:
            case Parquet:
            case Orc:
            case W3CLogFile:
                TransformationMethod transformationMethod = getTransform();
                return !StringUtils.isEmpty(this.columnName) && (!StringUtils.isEmpty(getPath())
                        || transformationMethod == TransformationMethod.SourceLineNumber
                        || transformationMethod == TransformationMethod.SourceLocation);
            case Avro:
            case ApacheAvro:
                return !StringUtils.isEmpty(this.columnName) &&
                        !StringUtils.isEmpty(getColumns());
            default:
                return false;
        }
    }
}

File: ingest-v2/src/main/kotlin/com/microsoft/azure/kusto/ingest/v2/common/models/mapping/ColumnMapping.kt
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
package com.microsoft.azure.kusto.ingest.v2.common.models.mapping

import com.microsoft.azure.kusto.ingest.v2.models.Format
import kotlinx.serialization.Serializable as KSerializable

@KSerializable
data class ColumnMapping(
    val columnName: String,
    val columnType: String,
    val properties: MutableMap<String, String> = mutableMapOf(),
) {

    fun setPath(path: String) {
        properties[MappingConstants.Path.name] = path
    }

    fun getPath(): String? = properties[MappingConstants.Path.name]

    fun setTransform(transform: TransformationMethod) {
        properties[MappingConstants.Transform.name] = transform.name
    }

    fun getTransform(): TransformationMethod? {
        val transform = properties[MappingConstants.Transform.name]
        return if (transform.isNullOrBlank()) {
            null
        } else {
            TransformationMethod.valueOf(transform)
        }
    }

    fun setOrdinal(ordinal: Int) {
        properties[MappingConstants.Ordinal.name] = ordinal.toString()
    }

    fun getOrdinal(): Int? {
        val ordinal = properties[MappingConstants.Ordinal.name]
        return if (ordinal.isNullOrBlank()) null else ordinal.toInt()
    }

    fun setConstantValue(constValue: String) {
        properties[MappingConstants.ConstValue.name] = constValue
    }

    fun getConstantValue(): String? =
        properties[MappingConstants.ConstValue.name]

    fun setField(field: String) {
        properties[MappingConstants.Field.name] = field
    }

    fun getField(): String? = properties[MappingConstants.Field.name]

    fun setColumns(columns: String) {
        properties[MappingConstants.Columns.name] = columns
    }

    fun getColumns(): String? = properties[MappingConstants.Columns.name]

    fun setStorageDataType(dataType: String) {
        properties[MappingConstants.StorageDataType.name] = dataType
    }

    fun getStorageDataType(): String? =
        properties[MappingConstants.StorageDataType.name]

    fun isValid(mappingKind: Format): Boolean {
        return when (mappingKind) {
            Format.csv,
            Format.sstream,
            -> columnName.isNotBlank()
            Format.json,
            Format.parquet,
            Format.orc,
            Format.w3clogfile,
            -> {
                val transformationMethod = getTransform()
                columnName.isNotBlank() &&
                    (
                        !getPath().isNullOrBlank() ||
                            transformationMethod ==
                            TransformationMethod.SourceLineNumber ||
                            transformationMethod ==
                            TransformationMethod.SourceLocation
                        )
            }
            Format.avro,
            Format.apacheavro,
            ->
                columnName.isNotBlank() && !getColumns().isNullOrBlank()
            else -> false
        }
    }
}
